libraries/AP_CANManager/AP_CANManager.cpp

```diff
@@ -424,13 +435,102 @@ void AP_CANManager::handle_can_frame(const mavlink_message_t &msg) const
 }
 
+/*
+  handle a CAN_FILTER_MODIFY packet
+ */
+void AP_CANManager::handle_can_filter_modify(const mavlink_message_t &msg)
+{
+    mavlink_can_filter_modify_t p;
+    mavlink_msg_can_filter_modify_decode(&msg, &p);
+    const int8_t bus = int8_t(p.bus)-1;
+    if (bus >= HAL_NUM_CAN_IFACES || hal.can[bus] == nullptr) {
+        return;
+    }
+    if (p.num_ids > ARRAY_SIZE(p.ids)) {
+        return;
+    }
+    uint16_t *new_ids = nullptr;
+    uint16_t num_new_ids = 0;
+    WITH_SEMAPHORE(can_forward.sem);
+
+    switch (p.operation) {
+    case CAN_FILTER_REPLACE: {
+        if (p.num_ids == 0) {
+            can_forward.num_filter_ids = 0;
+            delete[] can_forward.filter_ids;
+            can_forward.filter_ids = nullptr;
+            return;
+        }
+        new_ids = new uint16_t[p.num_ids];
+        if (new_ids != nullptr) {
+            num_new_ids = p.num_ids;
+            memcpy((void*)new_ids, (const void *)p.ids, p.num_ids*sizeof(uint16_t));
+            // sort the list, so we can bisection search
+            insertion_sort_uint16(new_ids, num_new_ids);
+        }
+        break;
+    }
+    case CAN_FILTER_ADD: {
+        new_ids = new uint16_t[can_forward.num_filter_ids+p.num_ids];
+        if (new_ids == nullptr) {
+            return;
+        }
+        memcpy(new_ids, can_forward.filter_ids, can_forward.num_filter_ids*sizeof(uint16_t));
+        memcpy(&new_ids[can_forward.num_filter_ids], p.ids, p.num_ids*sizeof(uint16_t));
+        insertion_sort_uint16(new_ids, can_forward.num_filter_ids+p.num_ids);
+        num_new_ids = remove_duplicates_uint16(new_ids, can_forward.num_filter_ids+p.num_ids);
+        break;
+    }
+    case CAN_FILTER_REMOVE: {
+        insertion_sort_uint16(p.ids, p.num_ids);
+        can_forward.num_filter_ids = remove_list_uint16(can_forward.filter_ids, can_forward.num_filter_ids,
+                                                        p.ids, p.num_ids);
+        break;
+    }
+    }
+    if (new_ids != nullptr) {
+        delete[] can_forward.filter_ids;
+        can_forward.filter_ids = new_ids;
+        can_forward.num_filter_ids = num_new_ids;
+    }
+}
+
 /*
   handler for CAN frames from the registered callback, sending frames
   out as CAN_FRAME messages
  */
 void AP_CANManager::can_frame_callback(uint8_t bus, const AP_HAL::CANFrame &frame)
 {
+    WITH_SEMAPHORE(can_forward.sem);
+    if (can_forward.frame_counter++ == 100) {
+        // check every 100 frames for disabling CAN_FRAME send
+        // we stop sending after 5s if the client stops
+        // sending MAV_CMD_CAN_FORWARD requests
+        if (AP_HAL::millis() - can_forward.last_callback_enable_ms > 5000) {
+            hal.can[bus]->register_frame_callback(nullptr);
+            return;
+        }
+        can_forward.frame_counter = 0;
+    }
+    WITH_SEMAPHORE(comm_chan_lock(can_forward.chan));
     if (HAVE_PAYLOAD_SPACE(can_forward.chan, CAN_FRAME)) {
+        if (can_forward.filter_ids != nullptr) {
```

Comment:
check if can_forward.num_filter_ids == 0?

Reply:
it is equivalent, but I've changed it to num_filter_ids. Makes no difference which one we check